### PR TITLE
Change the default mode of CachingOptimizer

### PR DESCRIPTION
**Pull request by odow**
Another way of fixing #1249.

The default of a `CachingOptimizer` is to initialize in `ATTACHED_OPTIMIZER` mode. This has two consequences:
* Solvers that do not support incremental building will throw an error at the first incremental call and reset the optimizer. 
* Solvers that do support incremental building will build incrementally, even if they have a fast `copy_to` available.

It seems like a much more sensible default is to initialize in `EMPTY_OPTIMIZER` mode. 
* Solvers that do not support incremental building no longer need to catch an error
* Solvers that do support incremental building will use the fast `copy_to` if available.

This is breaking (see the test change), so we may want to hold off until MOI 0.10, but it does need doing.

### CODE REVIEW

**Review by mlubin — APPROVED** [review]
Makes sense to me, pretty clean fix. Agreed it should be for MOI 0.10.

**Review by blegat — APPROVED** [review]
Looks good. I don't think it's breaking. We can release this in a patch release. It only affect the state in `AUTOMATIC` mode anyway

File: test/Utilities/cachingoptimizer.jl [hunk 1]
```diff
@@ -108,7 +108,11 @@ struct DummyConstraintAttribute <: MOI.AbstractConstraintAttribute end
...
 @testset "Mapping of variables" begin
     mock = MOIU.MockOptimizer(MOIU.UniversalFallback(MOIU.Model{Float64}()))
-    model = MOIU.CachingOptimizer(MOIU.UniversalFallback(MOIU.Model{Float64}()), mock)
+    model = MOIU.CachingOptimizer(
+        MOIU.UniversalFallback(MOIU.Model{Float64}()),
+        mock,
+    )
+    MOIU.attach_optimizer(model)
     x = MOI.add_variable(model)
     y = first(MOI.get(mock, MOI.ListOfVariableIndices()))
```
Comment (odow) [inline test/Utilities/cachingoptimizer.jl:117]: @blegat it's breaking here without an explicit `attach_optimizer`. People doing stuff with the backend of JuMP before `optimize!` may run into errors.

Comment (blegat) [reply test/Utilities/cachingoptimizer.jl:117]: Which error ? In `AUTOMATIC` mode, we automatically attach instead of erroring.

Comment (odow) [reply test/Utilities/cachingoptimizer.jl:117]: This example queries the inner optimizer of the cache directly.

Comment (blegat) [reply test/Utilities/cachingoptimizer.jl:117]: You shouldn't expect anything on the state of the inner optimizer in `AUTOMATIC` mode.
The only good reason users were accessing the inner optimizer was for optimizer attributes but that is not changed by this PR.

**Review by blegat — APPROVED** [review]

**Pull request merged**